File: sdk-python/synapse/policies/builtin.py
"""Four built-in MergePolicies covering the common cases."""
from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Optional

from .base import MergeAction, MergeDecision, MergePolicy, SynapseConflict

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# NoOp — proceed with a warning. Equivalent to v0.1 behavior.
# ---------------------------------------------------------------------------
class NoOpPolicy(MergePolicy):
    """Proceed silently after logging the conflict."""

    name = "no_op"

    async def resolve(self, handle, conflicts, proposed_action=None):
        return MergeAction(
            decision=MergeDecision.PROCEED,
            rationale=f"Proceeding through {len(conflicts)} conflict(s) (no_op policy).",
        )


# ---------------------------------------------------------------------------
# Abort — fail the intention with a SynapseConflict.
# ---------------------------------------------------------------------------
class AbortPolicy(MergePolicy):
    """Fail the intention. The caller's framework receives a clean
    ``SynapseConflict`` exception it can handle natively."""

    name = "abort"

    async def resolve(self, handle, conflicts, proposed_action=None):
        return MergeAction(
            decision=MergeDecision.ABORT,
            rationale=(
                f"Aborted: {len(conflicts)} other agent(s) hold conflicting "
                f"intentions on {handle.scope}. Caller should pivot."
            ),
        )


# ---------------------------------------------------------------------------
# Wait — block until prior conflicts resolve, then proceed.
# ---------------------------------------------------------------------------
class WaitPolicy(MergePolicy):
    """Block briefly then proceed. The intend() runtime polls the state
    graph for the conflicting intentions to flip to status='resolved'.
    """

    name = "wait"

    def __init__(self, timeout_ms: int = 5000):
        self.timeout_ms = timeout_ms

    async def resolve(self, handle, conflicts, proposed_action=None):
        return MergeAction(
            decision=MergeDecision.WAIT,
            wait_timeout_ms=self.timeout_ms,
            rationale=f"Waiting up to {self.timeout_ms}ms for {len(conflicts)} prior intention(s) to resolve.",
        )


# ---------------------------------------------------------------------------
# Redirect — re-emit with the other agent's recent work as context.
# ---------------------------------------------------------------------------
class RedirectPolicy(MergePolicy):
    """Default policy: surface the conflict to the caller as a structured
    rationale so the caller's framework / agent can re-prompt the LLM
    with the other agent's recent work in context.

    This policy doesn't itself call an LLM. It just packages the
    conflict information cleanly. It's the safest default for
    production: alerts loud, never silently merges, never blocks.
    """

    name = "redirect"

    async def resolve(self, handle, conflicts, proposed_action=None):
        others = sorted({
            ci.agent_id for c in conflicts
            for ci in (getattr(c, "conflicting_intentions", None) or [])
        }) or ["other agent(s)"]

        suggested = ", ".join(
            getattr(c, "suggested_resolution", "pivot") or "pivot" for c in conflicts
        )
        rationale = (
            f"Other agent(s) ({', '.join(others)}) hold overlapping "
            f"intention(s) on scope {handle.scope}. Suggested action: "
            f"{suggested}. Caller should pivot — re-prompt the LLM with "
            f"the other agent(s)' recent work in context, then re-invoke."
        )
        return MergeAction(
            decision=MergeDecision.PROCEED,
            rationale=rationale,
        )


# ---------------------------------------------------------------------------
# Auto-merge — opt-in: ask the user's BYO-LLM to merge two writes.
# ---------------------------------------------------------------------------
class AutoMergePolicy(MergePolicy):
    """Opt-in: ask the user's configured LLM (via ``synapse.set_llm``)
    to merge the conflicting writes and return a unified version.

    Requires:
      1. ``synapse.set_llm()`` has been called (or auto-detect succeeds)
      2. ``proposed_action`` contains the agent's planned content (e.g.
         ``{"path": "user.py", "content": "..."}``)
      3. The conflicting RESOLUTION's ``state_diff`` carries the other
         agent's content under a known key (``content``, ``output``, or
         ``output_preview``).

    On success, returns ``MergeAction(decision=MERGED, merged_action=...)``
    with merged content under the same key the caller used. On failure
    (no LLM configured, missing content, LLM error), falls back to
    REDIRECT semantics — never silently writes the wrong thing.
    """

    name = "auto_merge"

    def __init__(self, *, content_key: str = "content"):
        self.content_key = content_key

    async def resolve(self, handle, conflicts, proposed_action=None):
        from synapse.llm.config import get_internal_llm
        from synapse.policies.base import MergeAction, MergeDecision

        llm = get_internal_llm()
        if llm is None:
            logger.info(
                "auto_merge: no LLM configured (synapse.set_llm() unset). "
                "Falling back to redirect."
            )
            return MergeAction(
                decision=MergeDecision.PROCEED,
                rationale="auto_merge skipped (no LLM); use synapse.set_llm() to enable.",
            )

        if not proposed_action or self.content_key not in proposed_action:
            return MergeAction(
                decision=MergeDecision.PROCEED,
                rationale=(
                    f"auto_merge requires proposed_action[{self.content_key!r}] — "
                    f"caller didn't supply it. Falling back to redirect."
                ),
            )

        # Find ALL conflicting agents' recent content (not just the first).
        # Multi-agent collisions need a multi-way merge.
        priors = await _fetch_all_prior_content(handle, conflicts, self.content_key)
        if not priors:
            return MergeAction(
                decision=MergeDecision.PROCEED,
                rationale=(
                    "auto_merge: couldn't fetch prior agent's content "
                    "(state graph unavailable or no state_diff). Redirecting."
                ),
            )

        my_content = proposed_action[self.content_key]
        merged = await _llm_merge_multi(
            llm,
            priors=priors,
            my_agent=handle.agent_id,
            my_content=my_content,
            scope=handle.scope,
        )

        if not merged:
            return MergeAction(
                decision=MergeDecision.PROCEED,
                rationale="auto_merge: LLM returned empty merge. Falling back to redirect.",
            )

        merged_action = dict(proposed_action)
        merged_action[self.content_key] = merged
        prior_names = ", ".join(p["agent_id"] for p in priors)
        return MergeAction(
            decision=MergeDecision.MERGED,
            merged_action=merged_action,
            rationale=(
                f"Auto-merged {handle.agent_id}'s draft with {len(priors)} prior "
                f"agent(s) ({prior_names}) via LLM."
            ),
        )


async def _fetch_all_prior_content(
    handle, conflicts, content_key: str
) -> list[dict]:
    """Pull EVERY conflicting agent's recent content. Returns a list
    (possibly empty); each item has ``agent_id``, ``intention_id``, ``content``.
    """
    from synapse.intend import _get_or_init_runtime

    rt = _get_or_init_runtime()
    bus = rt.get("bus")
    if bus is None:
        return []

    handle_session_id = getattr(handle, "session_id", None)
    out: list[dict] = []
    seen_intents: set[str] = set()

    for c in conflicts:
        cis = getattr(c, "conflicting_intentions", None) or []
        for ci in cis:
            int_id = getattr(ci, "intention_id", None) or (ci.get("intention_id") if isinstance(ci, dict) else None)
            agent_id = getattr(ci, "agent_id", None) or (ci.get("agent_id") if isinstance(ci, dict) else "unknown")
            if not int_id or int_id in seen_intents:
                continue
            seen_intents.add(int_id)
            content = await _read_resolution_state_diff(rt, int_id, content_key, session_id=handle_session_id)
            logger.info(
                "auto_merge: prior intention=%s agent=%s -> content_len=%d",
                int_id, agent_id, len(content) if content else 0,
            )
            if content:
                out.append({"agent_id": str(agent_id), "intention_id": int_id, "content": content})
    return out


async def _fetch_prior_content(
    handle, conflicts, content_key: str
) -> Optional[dict]:
    """Pull the conflicting agent's RESOLUTION state_diff from the state
    graph (which holds the most recent finished writes).
    """
    from synapse.intend import _get_or_init_runtime

    rt = _get_or_init_runtime()
    bus = rt.get("bus")
    if bus is None:
        logger.info("auto_merge: rt has no bus (mode=%s)", rt.get("mode"))
        return None

    n_conflicts = len(conflicts)
    n_cis = sum(len(getattr(c, "conflicting_intentions", None) or []) for c in conflicts)
    logger.info("auto_merge: scanning %d conflict(s) with %d conflicting_intention(s)", n_conflicts, n_cis)

    # Use the IntentionHandle's session_id directly — stale cache entries
    # from earlier sessions can mislead the runtime peek.
    handle_session_id = getattr(handle, "session_id", None)

    # The conflicts come from the L2 router's CONFLICT envelope
    for c in conflicts:
        cis = getattr(c, "conflicting_intentions", None) or []
        for ci in cis:
            int_id = getattr(ci, "intention_id", None) or (ci.get("intention_id") if isinstance(ci, dict) else None)
            agent_id = getattr(ci, "agent_id", None) or (ci.get("agent_id") if isinstance(ci, dict) else "unknown")
            if not int_id:
                continue
            content = await _read_resolution_state_diff(rt, int_id, content_key, session_id=handle_session_id)
            logger.info("auto_merge: prior intention=%s agent=%s -> content_len=%d",
                        int_id, agent_id, len(content) if content else 0)
            if content:
                return {"agent_id": str(agent_id), "intention_id": int_id, "content": content}
    return None


async def _read_resolution_state_diff(
    rt: dict, intention_id: str, content_key: str, session_id: Optional[str] = None,
) -> Optional[str]:
    """Look up the RESOLUTION for ``intention_id`` on the session stream,
    return its state_diff[content_key] if present.
    """
    bus = rt.get("bus")
    sid = session_id or _peek_session_from_runtime(rt)
    if bus is None or sid is None:
        logger.info("auto_merge.lookup: bus=%s session_id=%s", bool(bus), sid)
        return None
    stream = f"synapse:session:{sid}:events"
    try:
        entries = await bus.redis.xrange(stream, count=500)
    except Exception as e:
        logger.warning("auto_merge.lookup: xrange failed: %s", e)
        return None
    n_entries = len(entries)
    n_resolutions = 0
    for _eid, fields in entries:
        try:
            env = json.loads(fields["e"])
        except Exception:
            continue
        if env.get("type") != "RESOLUTION":
            continue
        n_resolutions += 1
        payload = env.get("payload") or {}
        if payload.get("intention_id") != intention_id:
            continue
        sd = payload.get("state_diff") or {}
        logger.info("auto_merge.lookup: matched RESOLUTION for intent=%s; state_diff keys=%s",
                    intention_id, list(sd.keys()))
        for k in (content_key, "content", "output_preview", "output"):
            if k in sd and sd[k]:
                return str(sd[k])
    logger.info("auto_merge.lookup: scanned %d entries (%d RESOLUTIONs), no match for intent=%s",
                n_entries, n_resolutions, intention_id)
    return None


def _peek_session_from_runtime(rt: dict) -> Optional[str]:
    """Best-effort session lookup. ``intend()`` doesn't store the session
    on the runtime dict (each call passes its own), so we peek at the
    cached agents and use the first one's session.
    """
    agents = rt.get("agents") or {}
    for key in agents:
        # Cache key is f"{session_id}::{agent_id}"
        if "::" in key:
            return key.split("::", 1)[0]
    import os
    return os.environ.get("SYNAPSE_SESSION_ID")


async def _llm_merge_multi(
    llm,
    *,
    priors: list[dict],
    my_agent: str,
    my_content: str,
    scope: list[str],
) -> str:
    """Multi-way merge: combine my_content with N prior agents' content."""
    prior_blocks = "\n\n".join(
        f"Agent {p['agent_id']} wrote:\n```\n{p['content']}\n```"
        for p in priors
    )
    prompt = (
        f"Multiple AI agents wrote conflicting content for the same scope ({', '.join(scope)}).\n\n"
        f"PRIOR AGENT WRITES (in order, oldest first):\n\n"
        f"{prior_blocks}\n\n"
        f"NEW AGENT ({my_agent}) is about to write:\n```\n{my_content}\n```\n\n"
        f"Produce a single merged version that incorporates EVERY agent's intent.\n"
        f"  - Preserve fields/decisions from ALL agents — do not drop any contribution.\n"
        f"  - If two agents conflict semantically (e.g. different formulas), pick the\n"
        f"    one that looks more correct and add an inline comment.\n"
        f"  - Output only the merged content, no explanation, no markdown fences."
    )
    return await _llm_call_text(llm, prompt)


async def _llm_call_text(llm, prompt: str) -> str:
    """Generic LLM call helper supporting bridge + native adapters."""
    messages = [{"role": "user", "content": prompt}]

    # Path 1: bridge adapters
    try:
        if hasattr(llm, "generate"):
            text = await llm.generate(messages=messages, max_tokens=1500, temperature=0.0)
            if isinstance(text, str) and text.strip():
                return text.strip()
    except Exception as e:
        logger.warning("auto_merge: llm.generate failed (%s)", e)

    # Path 2: native Anthropic adapter
    client = getattr(llm, "_client", None)
    model = getattr(llm, "_model", None) or "claude-haiku-4-5-20251001"
    if client is not None and hasattr(client, "messages"):
        try:
            msg = await client.messages.create(
                model=model, max_tokens=1500, messages=messages,
            )
            blocks = msg.content if msg and getattr(msg, "content", None) else []
            text = blocks[0].text if blocks and hasattr(blocks[0], "text") else ""
            if text and text.strip():
                return text.strip()
        except Exception as e:
            logger.warning("auto_merge: anthropic client fallback failed (%s)", e)

    # Path 3: native OpenAI adapter
    if client is not None and hasattr(client, "chat") and hasattr(client.chat, "completions"):
        try:
            resp = await client.chat.completions.create(
                model=model, max_tokens=1500, messages=messages, temperature=0.0,
            )
            text = resp.choices[0].message.content if resp.choices else ""
            if text and text.strip():
                return text.strip()
        except Exception as e:
            logger.warning("auto_merge: openai client fallback failed (%s)", e)

    return ""


async def _llm_merge(
    llm,
    *,
    prior_agent: str,
    prior_content: str,
    my_agent: str,
    my_content: str,
    scope: list[str],
) -> str:
    """Call the user's LLM to produce a merged version.

    Tries ``llm.generate(messages=...)`` first (the bridge / cheap-variant
    path), then falls back to whatever the adapter exposes.
    """
    prompt = (
        f"Two AI agents both wrote content for the same scope ({', '.join(scope)}).\n\n"
        f"Agent A ({prior_agent}) wrote (just now, BEFORE you saw it):\n"
        f"```\n{prior_content}\n```\n\n"
        f"Agent B ({my_agent}) is about to write:\n"
        f"```\n{my_content}\n```\n\n"
        f"Produce a single merged version that incorporates both agents' "
        f"intent. Preserve fields/decisions from BOTH agents — do not drop "
        f"either's contribution. If they conflict semantically (e.g. "
        f"different formulas for the same field), prefer the one that "
        f"looks more correct and add an inline comment noting the choice.\n\n"
        f"Output ONLY the merged content, no explanation, no markdown fences."
    )
    messages = [{"role": "user", "content": prompt}]

    # Path 1: bridge adapters (LangChain / LiteLLM) — they expose .generate()
    try:
        if hasattr(llm, "generate"):
            text = await llm.generate(messages=messages, max_tokens=1500, temperature=0.0)
            if isinstance(text, str) and text.strip():
                return text.strip()
    except Exception as e:
        logger.warning("auto_merge: llm.generate failed (%s)", e)

    # Path 2: native Anthropic adapter — poke at the underlying client
    client = getattr(llm, "_client", None)
    model = getattr(llm, "_model", None) or "claude-haiku-4-5-20251001"
    if client is not None and hasattr(client, "messages"):
        try:
            msg = await client.messages.create(
                model=model,
                max_tokens=1500,
                messages=messages,
            )
            blocks = msg.content if msg and getattr(msg, "content", None) else []
            text = blocks[0].text if blocks and hasattr(blocks[0], "text") else ""
            if text and text.strip():
                return text.strip()
        except Exception as e:
            logger.warning("auto_merge: anthropic client fallback failed (%s)", e)

    # Path 3: native OpenAI adapter — chat.completions
    if client is not None and hasattr(client, "chat") and hasattr(client.chat, "completions"):
        try:
            resp = await client.chat.completions.create(
                model=model,
                max_tokens=1500,
                messages=messages,
                temperature=0.0,
            )
            text = resp.choices[0].message.content if resp.choices else ""
            if text and text.strip():
                return text.strip()
        except Exception as e:
            logger.warning("auto_merge: openai client fallback failed (%s)", e)

    return ""

File: sdk-typescript/src/policies/builtin.ts
/**
 * Five built-in MergePolicies covering the common cases.
 *
 * Ported from `sdk-python/synapse/policies/builtin.py`.
 *
 * NOTE on cross-module imports: `AutoMergePolicy.resolve()` reaches into
 * `../llm/config` and `../intend` which are owned by A1/A4 agents and may
 * not exist when this module is first compiled in isolation. We use
 * dynamic `import()` calls inside `resolve()` so the static graph stays
 * lean and test code can `vi.mock()` those module specifiers.
 */
import {
  MergeAction,
  MergeDecision,
  MergePolicy,
  type IntentionHandleLike,
} from "./base.js";
import type { Conflict, ConflictingIntention } from "../types.js";

// ---------------------------------------------------------------------------
// NoOp — proceed with a warning. Equivalent to v0.1 behavior.
// ---------------------------------------------------------------------------
export class NoOpPolicy extends MergePolicy {
  public override name = "no_op";

  async resolve(
    _handle: IntentionHandleLike,
    conflicts: Conflict[],
    _proposedAction?: Record<string, unknown>,
  ): Promise<MergeAction> {
    return {
      decision: MergeDecision.PROCEED,
      rationale: `Proceeding through ${conflicts.length} conflict(s) (no_op policy).`,
    };
  }
}

// ---------------------------------------------------------------------------
// Abort — fail the intention with a SynapseConflict.
// ---------------------------------------------------------------------------
export class AbortPolicy extends MergePolicy {
  public override name = "abort";

  async resolve(
    handle: IntentionHandleLike,
    conflicts: Conflict[],
    _proposedAction?: Record<string, unknown>,
  ): Promise<MergeAction> {
    return {
      decision: MergeDecision.ABORT,
      rationale:
        `Aborted: ${conflicts.length} other agent(s) hold conflicting ` +
        `intentions on ${JSON.stringify(handle.scope)}. Caller should pivot.`,
    };
  }
}

// ---------------------------------------------------------------------------
// Wait — block until prior conflicts resolve, then proceed.
// ---------------------------------------------------------------------------
export class WaitPolicy extends MergePolicy {
  public override name = "wait";
  public timeoutMs: number;

  constructor(timeoutMs = 5000) {
    super();
    this.timeoutMs = timeoutMs;
  }

  async resolve(
    _handle: IntentionHandleLike,
    conflicts: Conflict[],
    _proposedAction?: Record<string, unknown>,
  ): Promise<MergeAction> {
    return {
      decision: MergeDecision.WAIT,
      waitTimeoutMs: this.timeoutMs,
      rationale: `Waiting up to ${this.timeoutMs}ms for ${conflicts.length} prior intention(s) to resolve.`,
    };
  }
}

// ---------------------------------------------------------------------------
// Redirect — re-emit with the other agent's recent work as context.
// ---------------------------------------------------------------------------
export class RedirectPolicy extends MergePolicy {
  public override name = "redirect";

  async resolve(
    handle: IntentionHandleLike,
    conflicts: Conflict[],
    _proposedAction?: Record<string, unknown>,
  ): Promise<MergeAction> {
    const otherIds = new Set<string>();
    for (const c of conflicts) {
      const cis: ConflictingIntention[] = c.conflicting_intentions ?? [];
      for (const ci of cis) {
        if (ci.agent_id) otherIds.add(ci.agent_id);
      }
    }
    const others = otherIds.size > 0 ? Array.from(otherIds).sort() : ["other agent(s)"];

    const suggested = conflicts
      .map((c) => c.suggested_resolution || "pivot")
      .join(", ");

    const rationale =
      `Other agent(s) (${others.join(", ")}) hold overlapping ` +
      `intention(s) on scope ${JSON.stringify(handle.scope)}. Suggested action: ` +
      `${suggested}. Caller should pivot — re-prompt the LLM with ` +
      `the other agent(s)' recent work in context, then re-invoke.`;

    return {
      decision: MergeDecision.PROCEED,
      rationale,
    };
  }
}

// ---------------------------------------------------------------------------
// Auto-merge — opt-in: ask the user's BYO-LLM to merge two writes.
// ---------------------------------------------------------------------------

interface PriorContent {
  agentId: string;
  intentionId: string;
  content: string;
}

export class AutoMergePolicy extends MergePolicy {
  public override name = "auto_merge";
  public contentKey: string;

  constructor(opts: { contentKey?: string } = {}) {
    super();
    this.contentKey = opts.contentKey ?? "content";
  }

  async resolve(
    handle: IntentionHandleLike,
    conflicts: Conflict[],
    proposedAction?: Record<string, unknown>,
  ): Promise<MergeAction> {
    // Dynamic import to keep static graph lean and to make test-time
    // mocking via `vi.mock("../llm/config", ...)` straightforward.
    const llmConfig = (await import("../llm/config.js").catch(() => null)) as
      | { getInternalLlm?: () => unknown }
      | null;
    const getInternalLlm = llmConfig?.getInternalLlm;
    const llm = (getInternalLlm ? getInternalLlm() : null) as LlmLike | null;

    if (!llm) {
      return {
        decision: MergeDecision.PROCEED,
        rationale:
          "auto_merge skipped (no LLM); use synapse.setLlm() to enable.",
      };
    }

    if (!proposedAction || !(this.contentKey in proposedAction)) {
      return {
        decision: MergeDecision.PROCEED,
        rationale:
          `auto_merge requires proposedAction[${JSON.stringify(this.contentKey)}] — ` +
          `caller didn't supply it. Falling back to redirect.`,
      };
    }

    const priors = await fetchAllPriorContent(handle, conflicts, this.contentKey);
    if (priors.length === 0) {
      return {
        decision: MergeDecision.PROCEED,
        rationale:
          "auto_merge: couldn't fetch prior agent's content " +
          "(state graph unavailable or no state_diff). Redirecting.",
      };
    }

    const myContent = String(proposedAction[this.contentKey] ?? "");
    const merged = await llmMergeMulti(llm, {
      priors,
      myAgent: handle.agentId,
      myContent,
      scope: handle.scope,
    });

    if (!merged) {
      return {
        decision: MergeDecision.PROCEED,
        rationale: "auto_merge: LLM returned empty merge. Falling back to redirect.",
      };
    }

    const mergedAction = { ...proposedAction, [this.contentKey]: merged };
    const priorNames = priors.map((p) => p.agentId).join(", ");
    return {
      decision: MergeDecision.MERGED,
      mergedAction,
      rationale:
        `Auto-merged ${handle.agentId}'s draft with ${priors.length} prior ` +
        `agent(s) (${priorNames}) via LLM.`,
    };
  }
}

// ---------------------------------------------------------------------------
// Helpers (exported for tests)
// ---------------------------------------------------------------------------

async function fetchAllPriorContent(
  handle: IntentionHandleLike,
  conflicts: Conflict[],
  contentKey: string,
): Promise<PriorContent[]> {
  const intendMod = (await import("../intend.js").catch(() => null)) as unknown as
    | {
        getOrInitRuntime?: () => RuntimeLike;
        _getOrInitRuntime?: () => RuntimeLike;
      }
    | null;
  const getRt = intendMod?.getOrInitRuntime ?? intendMod?._getOrInitRuntime;
  if (!getRt) return [];

  const rt = getRt();
  if (!rt || !rt.bus) return [];

  const sessionId = handle.sessionId;
  const out: PriorContent[] = [];
  const seen = new Set<string>();

  for (const c of conflicts) {
    const cis = c.conflicting_intentions ?? [];
    for (const ci of cis) {
      const intId = ci.intention_id;
      const agentId = ci.agent_id ?? "unknown";
      if (!intId || seen.has(intId)) continue;
      seen.add(intId);
      const content = await readResolutionStateDiff(
        rt,
        intId,
        contentKey,
        sessionId,
      );
      if (content) {
        out.push({ agentId: String(agentId), intentionId: intId, content });
      }
    }
  }
  return out;
}

interface RuntimeLike {
  bus?: { redis?: { xrange: (stream: string, ...args: unknown[]) => Promise<Array<[string, Record<string, string>]>> } };
  agents?: Map<string, unknown> | Record<string, unknown>;
  mode?: string;
}

async function readResolutionStateDiff(
  rt: RuntimeLike,
  intentionId: string,
  contentKey: string,
  sessionId?: string,
): Promise<string | null> {
  const bus = rt.bus;
  const sid = sessionId || peekSessionFromRuntime(rt);
  if (!bus || !bus.redis || !sid) return null;

  const stream = `synapse:session:${sid}:events`;
  let entries: Array<[string, Record<string, string>]>;
  try {
    // ioredis xrange signature: redis.xrange(key, start, end, "COUNT", n)
    entries = await bus.redis.xrange(stream, "-", "+", "COUNT", 500);
  } catch {
    return null;
  }

  for (const [, fields] of entries) {
    const raw = fields["e"];
    if (!raw) continue;
    let env: Record<string, unknown>;
    try {
      env = JSON.parse(raw);
    } catch {
      continue;
    }
    if (env["type"] !== "RESOLUTION") continue;
    const payload = (env["payload"] as Record<string, unknown> | undefined) ?? {};
    if (payload["intention_id"] !== intentionId) continue;
    const sd = (payload["state_diff"] as Record<string, unknown> | undefined) ?? {};
    for (const k of [contentKey, "content", "output_preview", "output"]) {
      const v = sd[k];
      if (v !== undefined && v !== null && v !== "") {
        return String(v);
      }
    }
  }
  return null;
}

function peekSessionFromRuntime(rt: RuntimeLike): string | undefined {
  const agents = rt.agents;
  let keys: Iterable<string> = [];
  if (agents instanceof Map) {
    keys = agents.keys();
  } else if (agents && typeof agents === "object") {
    keys = Object.keys(agents);
  }
  for (const key of keys) {
    if (key.includes("::")) {
      return key.split("::", 2)[0];
    }
  }
  return process.env["SYNAPSE_SESSION_ID"];
}

interface LlmLike {
  generate?: (args: {
    messages: Array<{ role: string; content: string }>;
    max_tokens?: number;
    temperature?: number;
  }) => Promise<string>;
  _client?: {
    messages?: {
      create: (args: {
        model: string;
        max_tokens: number;
        messages: Array<{ role: string; content: string }>;
      }) => Promise<{ content?: Array<{ text?: string }> }>;
    };
    chat?: {
      completions?: {
        create: (args: {
          model: string;
          max_tokens: number;
          messages: Array<{ role: string; content: string }>;
          temperature?: number;
        }) => Promise<{ choices?: Array<{ message?: { content?: string } }> }>;
      };
    };
  };
  _model?: string;
}

async function llmMergeMulti(
  llm: LlmLike,
  args: {
    priors: PriorContent[];
    myAgent: string;
    myContent: string;
    scope: string[];
  },
): Promise<string> {
  const { priors, myAgent, myContent, scope } = args;
  const priorBlocks = priors
    .map((p) => `Agent ${p.agentId} wrote:\n\`\`\`\n${p.content}\n\`\`\``)
    .join("\n\n");
  const prompt =
    `Multiple AI agents wrote conflicting content for the same scope (${scope.join(", ")}).\n\n` +
    `PRIOR AGENT WRITES (in order, oldest first):\n\n` +
    `${priorBlocks}\n\n` +
    `NEW AGENT (${myAgent}) is about to write:\n\`\`\`\n${myContent}\n\`\`\`\n\n` +
    `Produce a single merged version that incorporates EVERY agent's intent.\n` +
    `  - Preserve fields/decisions from ALL agents — do not drop any contribution.\n` +
    `  - If two agents conflict semantically (e.g. different formulas), pick the\n` +
    `    one that looks more correct and add an inline comment.\n` +
    `  - Output only the merged content, no explanation, no markdown fences.`;
  return await llmCallText(llm, prompt);
}

async function llmCallText(llm: LlmLike, prompt: string): Promise<string> {
  const messages = [{ role: "user", content: prompt }];

  // Path 1: bridge adapters
  if (typeof llm.generate === "function") {
    try {
      const text = await llm.generate({
        messages,
        max_tokens: 1500,
        temperature: 0.0,
      });
      if (typeof text === "string" && text.trim()) return text.trim();
    } catch {
      // fall through to next path
    }
  }

  const client = llm._client;
  const model = llm._model ?? "claude-haiku-4-5-20251001";

  // Path 2: native Anthropic adapter
  if (client && client.messages && typeof client.messages.create === "function") {
    try {
      const msg = await client.messages.create({
        model,
        max_tokens: 1500,
        messages,
      });
      const blocks = msg?.content ?? [];
      const text = blocks[0]?.text ?? "";
      if (text && text.trim()) return text.trim();
    } catch {
      // fall through
    }
  }

  // Path 3: native OpenAI adapter
  if (
    client &&
    client.chat &&
    client.chat.completions &&
    typeof client.chat.completions.create === "function"
  ) {
    try {
      const resp = await client.chat.completions.create({
        model,
        max_tokens: 1500,
        messages,
        temperature: 0.0,
      });
      const text = resp?.choices?.[0]?.message?.content ?? "";
      if (text && text.trim()) return text.trim();
    } catch {
      // fall through
    }
  }

  return "";
}
